File: firmware/esp32-dualeye/src/boot_animation.cpp
#include <Arduino.h>
#include <math.h>

#include "boot_animation.h"
#include "config.h"

#if ENABLE_DISPLAY_BOOT_ANIMATION
#include <TFT_eSPI.h>

static TFT_eSPI tft = TFT_eSPI();

static uint16_t c565(uint8_t r, uint8_t g, uint8_t b) {
  return tft.color565(r, g, b);
}

static int scaled(float value, int base) {
  return (int)(value * (float)base);
}

static void glowCircle(int x, int y, int r, uint16_t core, uint16_t mid, uint16_t outer, float pulse) {
  int bloom = max(2, (int)(r * (0.8f + pulse * 1.2f)));
  tft.fillCircle(x, y, r + bloom, outer);
  tft.fillCircle(x, y, r + max(1, bloom / 2), mid);
  tft.fillCircle(x, y, r, core);
  tft.fillCircle(x + max(1, r / 3), y - max(1, r / 3), max(1, r / 5), TFT_WHITE);
}

static void drawSegmentedProgress(float progress) {
  const int w = tft.width();
  const int h = tft.height();
  const int barW = max(120, (int)(w * 0.70f));
  const int barH = max(6, (int)(h * 0.035f));
  const int x0 = (w - barW) / 2;
  const int y0 = h - max(42, (int)(h * 0.16f));
  const int segmentGap = 4;
  const int segmentW = 7;
  const int count = max(8, barW / (segmentW + segmentGap));
  const int lit = min(count, max(0, (int)(progress * (float)count)));

  const uint16_t dim = c565(24, 26, 34);
  const uint16_t magenta = c565(235, 70, 255);
  const uint16_t blue = c565(65, 170, 255);

  tft.drawLine(x0 - 14, y0 + barH / 2, x0 - 4, y0 + barH / 2, magenta);
  tft.drawLine(x0 + barW + 4, y0 + barH / 2, x0 + barW + 14, y0 + barH / 2, blue);

  for (int i = 0; i < count; ++i) {
    int x = x0 + i * (segmentW + segmentGap);
    uint16_t color = dim;
    if (i < lit) {
      color = (i < count / 2) ? magenta : blue;
    }
    tft.fillRoundRect(x, y0, segmentW, barH, 2, color);
  }
}

static void drawBootFrame(float progress, float pulse) {
  tft.fillScreen(TFT_BLACK);

  const int w = tft.width();
  const int h = tft.height();
  const int base = min(w, h);
  const int cx = w / 2;
  const int cy = scaled(0.36f, h);

  const int headR = scaled(0.265f, base);
  const int earR = scaled(0.135f, base);
  const int eyeR = scaled(0.047f, base) + (int)(pulse * scaled(0.018f, base));
  const int eyeDX = scaled(0.135f, base);
  const int eyeDY = -scaled(0.035f, base);

  const uint16_t face = c565(7, 8, 12);
  const uint16_t edge = c565(28, 31, 40);
  const uint16_t nose = c565(52, 55, 62);
  const uint16_t noseEdge = c565(112, 116, 126);
  const uint16_t tooth = c565(190, 192, 205);
  const uint16_t purple = c565(230, 60, 255);
  const uint16_t purpleDim = c565(65, 12, 95);
  const uint16_t blue = c565(60, 165, 255);
  const uint16_t blueDim = c565(8, 45, 110);
  const uint16_t textGray = c565(116, 118, 130);

  // Ears and face silhouette.
  tft.fillCircle(cx - scaled(0.31f, base), cy - scaled(0.22f, base), earR, face);
  tft.fillCircle(cx + scaled(0.31f, base), cy - scaled(0.22f, base), earR, face);
  tft.drawCircle(cx - scaled(0.31f, base), cy - scaled(0.22f, base), earR, edge);
  tft.drawCircle(cx + scaled(0.31f, base), cy - scaled(0.22f, base), earR, edge);
  tft.fillCircle(cx, cy, headR, face);
  tft.drawCircle(cx, cy, headR, edge);

  // Angry brow slashes.
  tft.drawLine(cx - scaled(0.20f, base), cy - scaled(0.12f, base), cx - scaled(0.04f, base), cy - scaled(0.025f, base), edge);
  tft.drawLine(cx + scaled(0.20f, base), cy - scaled(0.12f, base), cx + scaled(0.04f, base), cy - scaled(0.025f, base), edge);
  tft.drawLine(cx - scaled(0.20f, base), cy - scaled(0.11f, base), cx - scaled(0.04f, base), cy - scaled(0.015f, base), edge);
  tft.drawLine(cx + scaled(0.20f, base), cy - scaled(0.11f, base), cx + scaled(0.04f, base), cy - scaled(0.015f, base), edge);

  // Pulsing eyes: left purple, right true blue.
  glowCircle(cx - eyeDX, cy + eyeDY, eyeR, c565(255, 226, 255), purple, purpleDim, pulse);
  glowCircle(cx + eyeDX, cy + eyeDY, eyeR, c565(232, 246, 255), blue, blueDim, pulse);

  // Nose.
  const int noseW = scaled(0.15f, base);
  const int noseH = scaled(0.21f, base);
  tft.fillRoundRect(cx - noseW / 2, cy + scaled(0.02f, base), noseW, noseH, max(6, noseW / 3), nose);
  tft.drawRoundRect(cx - noseW / 2, cy + scaled(0.02f, base), noseW, noseH, max(6, noseW / 3), noseEdge);

  // Fangs / mouth.
  const int mouthY = cy + scaled(0.28f, base);
  tft.fillTriangle(cx - scaled(0.08f, base), mouthY, cx - scaled(0.03f, base), mouthY - scaled(0.035f, base), cx - scaled(0.01f, base), mouthY + scaled(0.045f, base), tooth);
  tft.fillTriangle(cx + scaled(0.08f, base), mouthY, cx + scaled(0.03f, base), mouthY - scaled(0.035f, base), cx + scaled(0.01f, base), mouthY + scaled(0.045f, base), tooth);
  tft.drawLine(cx - scaled(0.05f, base), mouthY + scaled(0.055f, base), cx + scaled(0.05f, base), mouthY + scaled(0.055f, base), edge);

  // Title.
  const int titleY = h - max(72, scaled(0.26f, h));
  const int textSize = (base >= 280) ? 2 : 1;
  tft.setTextDatum(MC_DATUM);
  tft.setTextSize(textSize);
  tft.setTextColor(purple, TFT_BLACK);
  tft.drawString("KoalaByte", cx - scaled(0.18f, base), titleY);
  tft.setTextColor(blue, TFT_BLACK);
  tft.drawString("Blue", cx + scaled(0.24f, base), titleY);

  drawSegmentedProgress(progress);

  tft.setTextSize(1);
  tft.setTextColor(textGray, TFT_BLACK);
  tft.drawString("BOOTING...", cx, h - max(16, scaled(0.07f, h)));
}

void setupDisplay() {
  tft.init();
  tft.setRotation(DISPLAY_ROTATION);
  tft.fillScreen(TFT_BLACK);
}

void runBootAnimation() {
  const uint32_t totalMs = BOOT_ANIMATION_TOTAL_MS;
  const uint32_t frameMs = BOOT_ANIMATION_FRAME_MS;
  const uint32_t startMs = millis();

  while ((millis() - startMs) < totalMs) {
    const uint32_t elapsed = millis() - startMs;
    float progress = (float)elapsed / (float)totalMs;
    if (progress > 1.0f) progress = 1.0f;

    // Smooth pulse from dim to bright and back while the progress bar advances.
    const float pulse = 0.5f + 0.5f * sinf(((float)elapsed / 220.0f) * 2.0f * PI);
    drawBootFrame(progress, pulse);
    delay(frameMs);
  }

  drawBootFrame(1.0f, 0.55f);
  delay(140);
}

#else

void setupDisplay() {}
void runBootAnimation() {}

#endif

File: pi-companion/koalablue/boot_animation.py
from __future__ import annotations

import math
import os
import time
from dataclasses import dataclass
from typing import Optional, Tuple


Color = Tuple[int, int, int]


@dataclass
class BootAnimationConfig:
    duration_seconds: float = 3.0
    fps: int = 30
    fullscreen: bool = True
    width: int = 800
    height: int = 480
    title: str = "KoalaByte Blue"


class BootSplashUnavailable(RuntimeError):
    pass


def _import_pygame():
    try:
        import pygame  # type: ignore
    except Exception as exc:  # pragma: no cover - depends on Pi display environment
        raise BootSplashUnavailable(f"pygame unavailable: {exc}") from exc
    return pygame


class KoalaByteBootAnimation:
    """Fullscreen Pi companion boot splash.

    Draws a procedural KoalaByte Blue boot screen with pulsing eyes and an
    advancing progress bar. This intentionally avoids binary image assets so the
    splash remains easy to version, install, and tune on the Pi.
    """

    def __init__(self, config: Optional[BootAnimationConfig] = None) -> None:
        self.config = config or BootAnimationConfig()
        self.pygame = _import_pygame()
        self.screen = None
        self.clock = None
        self.font_title = None
        self.font_small = None

    def setup(self) -> None:
        pygame = self.pygame
        os.environ.setdefault("PYGAME_HIDE_SUPPORT_PROMPT", "1")
        pygame.init()
        flags = pygame.FULLSCREEN if self.config.fullscreen else 0
        if self.config.fullscreen:
            self.screen = pygame.display.set_mode((0, 0), flags)
        else:
            self.screen = pygame.display.set_mode((self.config.width, self.config.height), flags)
        pygame.display.set_caption("KoalaByte Blue Boot")
        self.clock = pygame.time.Clock()
        w, h = self.screen.get_size()
        title_size = max(28, min(72, int(w * 0.058)))
        small_size = max(13, min(24, int(w * 0.018)))
        self.font_title = pygame.font.SysFont("dejavusansmono", title_size, bold=True)
        self.font_small = pygame.font.SysFont("dejavusansmono", small_size)

    def run(self) -> int:
        self.setup()
        pygame = self.pygame
        start = time.monotonic()
        total = max(0.25, self.config.duration_seconds)

        while True:
            elapsed = time.monotonic() - start
            progress = min(1.0, elapsed / total)
            pulse = 0.5 + 0.5 * math.sin(elapsed * math.tau * 2.8)
            self.draw(progress, pulse)
            pygame.display.flip()

            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    return 0
                if event.type == pygame.KEYDOWN and event.key in {pygame.K_ESCAPE, pygame.K_q}:
                    return 0

            if progress >= 1.0:
                time.sleep(0.18)
                return 0
            self.clock.tick(self.config.fps)

    def draw(self, progress: float, pulse: float) -> None:
        pygame = self.pygame
        screen = self.screen
        assert screen is not None
        w, h = screen.get_size()
        base = min(w, h)
        cx = w // 2
        cy = int(h * 0.36)

        black: Color = (0, 0, 0)
        face: Color = (6, 8, 12)
        edge: Color = (30, 34, 44)
        nose: Color = (54, 57, 65)
        nose_edge: Color = (118, 120, 132)
        tooth: Color = (185, 188, 202)
        purple: Color = (235, 70, 255)
        purple_dim: Color = (72, 8, 105)
        blue: Color = (70, 170, 255)
        blue_dim: Color = (6, 40, 115)
        text_gray: Color = (120, 122, 136)

        screen.fill(black)

        def sc(v: float) -> int:
            return int(v * base)

        # Koala silhouette.
        ear_r = sc(0.15)
        head_r = sc(0.27)
        pygame.draw.circle(screen, face, (cx - sc(0.32), cy - sc(0.21)), ear_r)
        pygame.draw.circle(screen, face, (cx + sc(0.32), cy - sc(0.21)), ear_r)
        pygame.draw.circle(screen, edge, (cx - sc(0.32), cy - sc(0.21)), ear_r, max(1, sc(0.006)))
        pygame.draw.circle(screen, edge, (cx + sc(0.32), cy - sc(0.21)), ear_r, max(1, sc(0.006)))
        pygame.draw.circle(screen, face, (cx, cy), head_r)
        pygame.draw.circle(screen, edge, (cx, cy), head_r, max(1, sc(0.006)))

        # Brows.
        brow_w = max(2, sc(0.012))
        pygame.draw.line(screen, edge, (cx - sc(0.22), cy - sc(0.12)), (cx - sc(0.04), cy - sc(0.02)), brow_w)
        pygame.draw.line(screen, edge, (cx + sc(0.22), cy - sc(0.12)), (cx + sc(0.04), cy - sc(0.02)), brow_w)

        # Eye glow, left purple and right actually blue.
        eye_r = sc(0.047) + int(pulse * sc(0.017))
        self._glow_circle((cx - sc(0.135), cy - sc(0.035)), eye_r, (255, 226, 255), purple, purple_dim, pulse)
        self._glow_circle((cx + sc(0.135), cy - sc(0.035)), eye_r, (232, 246, 255), blue, blue_dim, pulse)

        # Nose and teeth.
        nose_rect = pygame.Rect(0, 0, sc(0.15), sc(0.22))
        nose_rect.center = (cx, cy + sc(0.11))
        pygame.draw.rect(screen, nose, nose_rect, border_radius=max(8, sc(0.04)))
        pygame.draw.rect(screen, nose_edge, nose_rect, max(1, sc(0.005)), border_radius=max(8, sc(0.04)))

        mouth_y = cy + sc(0.29)
        pygame.draw.polygon(screen, tooth, [(cx - sc(0.09), mouth_y), (cx - sc(0.035), mouth_y - sc(0.04)), (cx - sc(0.01), mouth_y + sc(0.05))])
        pygame.draw.polygon(screen, tooth, [(cx + sc(0.09), mouth_y), (cx + sc(0.035), mouth_y - sc(0.04)), (cx + sc(0.01), mouth_y + sc(0.05))])
        pygame.draw.line(screen, edge, (cx - sc(0.06), mouth_y + sc(0.06)), (cx + sc(0.06), mouth_y + sc(0.06)), max(1, sc(0.006)))

        # Title: KoalaByte purple, Blue true blue.
        title_y = int(h * 0.78)
        self._draw_centered_split_text(cx, title_y, "KoalaByte", purple, "Blue", blue)

        self._draw_progress_bar(progress, purple, blue)

        small = self.font_small.render("B O O T I N G . . .", True, text_gray)
        screen.blit(small, small.get_rect(center=(cx, int(h * 0.94))))

    def _glow_circle(self, center: Tuple[int, int], radius: int, core: Color, mid: Color, outer: Color, pulse: float) -> None:
        pygame = self.pygame
        screen = self.screen
        assert screen is not None
        bloom = int(radius * (1.2 + pulse * 1.8))
        pygame.draw.circle(screen, outer, center, radius + bloom)
        pygame.draw.circle(screen, mid, center, radius + bloom // 2)
        pygame.draw.circle(screen, core, center, radius)
        pygame.draw.circle(screen, (255, 255, 255), (center[0] + radius // 3, center[1] - radius // 3), max(2, radius // 6))

    def _draw_centered_split_text(self, cx: int, y: int, left: str, left_color: Color, right: str, right_color: Color) -> None:
        screen = self.screen
        assert screen is not None
        left_surf = self.font_title.render(left, True, left_color)
        gap = max(14, int(screen.get_width() * 0.018))
        right_surf = self.font_title.render(right, True, right_color)
        total_w = left_surf.get_width() + gap + right_surf.get_width()
        x = cx - total_w // 2
        screen.blit(left_surf, (x, y - left_surf.get_height() // 2))
        screen.blit(right_surf, (x + left_surf.get_width() + gap, y - right_surf.get_height() // 2))

    def _draw_progress_bar(self, progress: float, purple: Color, blue: Color) -> None:
        pygame = self.pygame
        screen = self.screen
        assert screen is not None
        w, h = screen.get_size()
        cx = w // 2
        bar_w = int(w * 0.60)
        y = int(h * 0.865)
        seg_w = max(8, int(w * 0.012))
        gap = max(4, int(w * 0.006))
        count = max(10, bar_w // (seg_w + gap))
        start_x = cx - ((count * seg_w + (count - 1) * gap) // 2)
        lit = int(count * progress)
        dim: Color = (30, 32, 42)

        pygame.draw.line(screen, purple, (start_x - 50, y + seg_w // 2), (start_x - 10, y + seg_w // 2), 1)
        pygame.draw.line(screen, blue, (start_x + count * (seg_w + gap) + 10, y + seg_w // 2), (start_x + count * (seg_w + gap) + 50, y + seg_w // 2), 1)

        for i in range(count):
            color = dim
            if i < lit:
                color = purple if i < count // 2 else blue
            rect = pygame.Rect(start_x + i * (seg_w + gap), y, seg_w, max(4, seg_w // 3))
            pygame.draw.rect(screen, color, rect, border_radius=max(2, rect.height // 2))


def run_boot_splash(config: Optional[BootAnimationConfig] = None) -> int:
    return KoalaByteBootAnimation(config).run()


if __name__ == "__main__":
    raise SystemExit(run_boot_splash())
